Not all possible namespace groups were added

dgedge33 opened the pull request:
Considering the TestCaseProject, if we rename the root namespace TestDoc to a composed name (SHFB.TestDoc for example), no more namespace group is added.

EWSoftware commented:
If all TestDoc namespaces are renamed SHFB.TestDoc, then this fix causes previously grouped namespaces such as TestDoc.Generics to remain ungrouped.  I think what really needs to happen here is to adjust maxParts based on the namespace with the minimum number of parts.  If said namespace has the same or more parts, maxParts needs to be increased accordingly to account for the new base size.

EWSoftware commented:
The following change fixes the issue and retains the proper grouping:

``` C#
int partCount;

// If the shortest namespace has more than one part, increase the maximum number of parts
// to account for the base length.
if(namespaces.Count() != 0)
    maxParts += namespaces.Min(n => n.Split('.').Length) - 1;

// This serves as the root group.  If a project element is present in the reflection file, its
// list of elements will be replaced by the list in this group.  If not, this group is written to
// the file to serve as a place holder for the TOC XSL transformation.
groups.Add(String.Empty, new NamespaceGroup { Namespace = String.Empty });
```

dgedge33 commented:
Hi Eric,

Oups, I miss the regression on TestDoc.Generics namespace.
If I fix this regression in my change, the output differs with the change you propose:

-          With your change we will have two groups added: G:SHFB.TestDoc.HierarchicalToc and G:SHFB.TestDoc.Generics

-          With my change we will have three groups added:

-           Adding G:SHFB.TestDoc.HierarchicalToc.WoodDreams.Reporting

-           Adding G:SHFB.TestDoc.HierarchicalToc

-           Adding G:SHFB.TestDoc.Generics

-           Added 3 namespace group entries

I think that the G:SHFB.TestDoc.HierarchicalToc.WoodDreams.Reporting is a legal namespace group.
What do you think about ?

Thanks.

From: Eric Woodruff [mailto:notifications@github.com]
Sent: mercredi 29 avril 2015 03:39
To: EWSoftware/SHFB
Cc: Lague, Jerome
Subject: Re: [SHFB] Not all possible namespace groups were added (#80)


The following change fixes the issue and retains the proper grouping:

int partCount;



// If the shortest namespace has more than one part, increase the maximum number of parts

// to account for the base length.

if(namespaces.Count() != 0)

    maxParts += namespaces.Min(n => n.Split('.').Length) - 1;



// This serves as the root group.  If a project element is present in the reflection file, its

// list of elements will be replaced by the list in this group.  If not, this group is written to

// the file to serve as a place holder for the TOC XSL transformation.

groups.Add(String.Empty, new NamespaceGroup { Namespace = String.Empty });

—
Reply to this email directly or view it on GitHub<https://github.com/EWSoftware/SHFB/pull/80#issuecomment-97281062>.

dgedge33 commented:
Sorry for my quick answer. 
Increasing the max group parts in the project properties did the trick.
